### ISSUE TITLE
Unexpected effects of `Route.Add`, `Route.Get`, `Route.Delete`

### ISSUE THREAD

**Issue by llorllale** [jsimonetti/rtnetlink#207]
<details><summary>Test Case</summary>

```go
func TestRoutes(t *testing.T) {
	lo, err := net.InterfaceByName("lo")
	require.NoError(t, err)

	_, dst, err := net.ParseCIDR("176.125.239.0/24")
	require.NoError(t, err)

	conn, err := rtnetlink.Dial(nil)
	require.NoError(t, err)

	t.Cleanup(func() {
		require.NoError(t, conn.Close())
	})

	ones, _ := dst.Mask.Size()

	routeMsg := &rtnetlink.RouteMessage{
		Family:    unix.AF_INET,
		Table:     unix.RT_TABLE_MAIN,
		Protocol:  unix.RTPROT_BOOT,
		Scope:     unix.RT_SCOPE_LINK,
		Type:      unix.RTN_UNICAST,
		DstLength: uint8(ones),
		Attributes: rtnetlink.RouteAttributes{
			Dst:      dst.IP,
			OutIface: uint32(lo.Index),
		},
	}

	err = conn.Route.Add(routeMsg)
	require.NoError(t, err)

	out, err := exec.Command("ip", "route", "show", "table", "main").CombinedOutput()
	require.NoError(t, err)

	fmt.Println(string(out))

	routes, err := conn.Route.Get(routeMsg)
	require.NoError(t, err)
	require.Len(t, routes, 1)
	route := routes[0]
	assert.Equal(t, int(routeMsg.DstLength), int(route.DstLength))

	err = conn.Route.Delete(routeMsg)
	require.NoError(t, err)

	out, err = exec.Command("ip", "route", "show", "table", "main").CombinedOutput()
	require.NoError(t, err)

	fmt.Println(string(out))

	routes, err = conn.Route.Get(routeMsg)
	require.NoError(t, err)
	assert.Empty(t, routes)
}
```
</details>

<details><summary>Test Output</summary>

```
=== RUN   TestRoutes
default via 10.88.111.254 dev wlp0s20f3 proto dhcp metric 600 
10.88.111.0/24 dev wlp0s20f3 proto kernel scope link src 10.88.111.17 metric 600 
169.254.0.0/16 dev wlp0s20f3 scope link metric 1000 
176.125.239.0/24 dev lo scope link 

    main_test.go:77: 
        	Error Trace:	/home/llorllale/GolandProjects/awesomeProject1/main_test.go:77
        	Error:      	Not equal: 
        	            	expected: 24
        	            	actual  : 32
        	Test:       	TestRoutes
default via 10.88.111.254 dev wlp0s20f3 proto dhcp metric 600 
10.88.111.0/24 dev wlp0s20f3 proto kernel scope link src 10.88.111.17 metric 600 
169.254.0.0/16 dev wlp0s20f3 scope link metric 1000 

    main_test.go:89: 
        	Error Trace:	/home/llorllale/GolandProjects/awesomeProject1/main_test.go:89
        	Error:      	Should be empty, but was [{2 32 0 0 254 0 0 1 2147484160 {176.125.239.0 10.88.111.17 <nil> 1 0 254 0 <nil> <nil> []}}]
        	Test:       	TestRoutes
--- FAIL: TestRoutes (0.03s)
```
</details>

TLDR:
* Conn.Route.Add adds the route with the expected `24` mask (as shown by first output of `ip route`)
* Conn.Route.Get returns a different mask
* Conn.Route.Del deletes the route (as shown by second output of `ip route`)
* Conn.Route.Get will **still** return the route, still with the wrong mask

What could be the issue here?

**Comment by jsimonetti** [comment]
Hi,

Redoing your tests with just the `iproute` command results in a similar error:
```bash
# sudo ip ro add 176.125.239.0/24 dev lo
# ip ro ls dev lo
176.125.239.0/24 scope link
# ip ro get 176.125.239.0/24 dev lo
Warning: /24 as prefix is invalid, only /32 (or none) is supported.
176.125.239.0 dev lo src x.x.x.x uid 1000
    cache <local>
```

I have to admit, I am not entirely sure why that happens, but I think the kernel is requiring an address and not a prefix in `route.Get`.
The manual of `iproute` also mentions this in their usage: `ip route get [ ROUTE_GET_FLAGS ] ADDRESS`
It specifically says `ADDRESS` and not `ROUTE` in the case of add: `ip route { add | del | change | append | replace } ROUTE`

If you add the `fibmatch` flag you get the result you are looking for:
```bash
# ip ro get fibmatch 176.125.239.1 dev lo
176.125.239.0/24 dev lo scope link
```

If you set your message accordingly, does that fix your issue?
`routeMsg.Flags = 0x2000`

**Comment by llorllale** [comment]
Hi,

> This seems to me that `route.Get` queries the kernel route cache, and not the FIB.

I've suspected that since I saw this [answer on SO](https://stackoverflow.com/a/15135509) - haven't yet confirmed it.

> If you set your message accordingly, does that fix your issue?
> `routeMsg.Flags = 0x1000|0x2000`

It does!

<details><summary>New Test Output</summary>

```
=== RUN   TestRoutes
default via 10.88.111.254 dev wlp0s20f3 proto dhcp metric 600 
10.88.111.0/24 dev wlp0s20f3 proto kernel scope link src 10.88.111.17 metric 600 
169.254.0.0/16 dev wlp0s20f3 scope link metric 1000 
176.125.239.0/24 dev lo scope link 

default via 10.88.111.254 dev wlp0s20f3 proto dhcp metric 600 
10.88.111.0/24 dev wlp0s20f3 proto kernel scope link src 10.88.111.17 metric 600 
169.254.0.0/16 dev wlp0s20f3 scope link metric 1000 

    main_test.go:86: 
        	Error Trace:	/home/llorllale/GolandProjects/awesomeProject1/main_test.go:86
        	Error:      	Received unexpected error:
        	            	netlink receive: no route to host
        	Test:       	TestRoutes
--- FAIL: TestRoutes (0.02s)
```
</details>

The test now fails reasonably on `conn.Route.Get` after having first deleted the route.

`0x1000` and `0x2000` correspond to `unix.RTM_F_LOOKUP_TABLE` and `unix.RTM_F_FIB_MATCH` respectively.

I'm having a hard time finding docs for these two flags aside for some patch notes [here](https://lkml.kernel.org/netdev/1495734160-47659-7-git-send-email-roopa@cumulusnetworks.com/) and [there](https://patchwork.ozlabs.org/project/netdev/patch/1442859588-14831-1-git-send-email-dsa@cumulusnetworks.com/) that assume the reader has the context to understand them.

**Comment by llorllale** [comment]
I guess I'll close this since I think your suggestion fixes the problem.

Do let me know if you know of any docs or whatever that can explain those two flags.

Thanks

**Comment by jsimonetti** [comment]
Unfortunately these kernel api's are very old and poorly ndocumented. When i look for authoritive information I usually look at how `iproute` handles stuff..
